doc/arch/008-screen-architecture-v3.md

```diff
@@ -0,0 +1,759 @@
+    whenever(settingsRepository.setCurrentLanguage(changeToLanguage)).doReturn(Completable.complete())
+
+    // when
+    testCase.dispatch(UpdateCurrentLanguageEffect(changeToLanguage))
+
+    // then
+    testCase.assertOutgoingEvents(CurrentLanguageChangedEvent)
+  }
+
+  @Test
+  fun `when the go back to previous screen effect is received, the go back ui action must be invoked`() {
+    // when
+    testCase.dispatch(GoBack)
+
+    // then
+    testCase.assertNoOutgoingEvents()
+    verify(uiActions).goBackToPreviousScreen()
+    verifyNoMoreInteractions(uiActions)
+  }
+
+  @Test
+  fun `when the restart activity effect is received, the restart activity ui action must be invoked`() {
+    // given
+    testCase.dispatch(RestartActivity)
+
+    // when
+    testCase.assertNoOutgoingEvents()
+
+    // then
+    verify(uiActions).restartActivity()
+    verifyNoMoreInteractions(uiActions)
+  }
+}
+```
+
+### Goals Review
+
+#### Fulfill all the requirements of the [v2 architecture](./007-screen-architecture-v2.md#goals)
+
+- **Separation of presentation and business logic:** This is satisfied by Mobius since it enables us to separate concerns at an even more granular level than before.
+- **Making UI state explicit and support state restoration:** This is handled by the `MobiusDelegate` classes for us and enforced by it since the class expects the `Model` to be `Parcelable` by default.
+
+#### Separation of concerns and testability
+We have three core components (`Update`, `EffectHandler`, and `UiRenderer`), which are responsible for discrete parts of the system. These are smaller and more focused, thereby making them easier to test and maintain as well.
+
+#### Migrating from the older architecture
+Unlike the [v2 architecture](./007-screen-architecture-v2.md#migrating-from-the-older-architecture), the v3 architecture is a much larger change from what we used to do earlier. As a consequence, migrating to the newer architecture is a much more involved process that requires us to follow deliberate, measured steps. The migration process is out of scope for this document, and is detailed in [this document](../migrating-to-mobius.md)
+
+## Consequences
+- In legacy controllers, events were cached and replayed as soon as the screen was inflated without waiting for it to be attached to the view hierarchy. In the v3 architecture, any events which are forwarded to the binding before the screen is attached to the view hierarchy, will be lost and ignored. This is the correct way to implement the events, but some of the older screen controllers which were dependent on that behaviour might take more effort to migrate to the current architecture.
+- Migrating to the v3 architecture from the v1 architecture is a process that takes some time to understand and get used to. Until this migration completes, new developers will need to be onboarded onto both the v1 and v3 architectures so that the codebase can be maintained.
```

Comment:
We should think and add more `Consequences` since this was a huge decision which has to have quite a bit of impact to justify why we chose it. We can discuss this. 
This could even mean that we will have to figure out a way to measure some of the goals we set in the beginning.